### PR TITLE
Run tasks upon Eclipse synchronization

### PR DESCRIPTION
**Pull request by donat**
This PR implements the Gradle side of the [Run tasks upon synchronization](https://github.com/eclipse/buildship/issues/265) Buildship story. It contributes a new `eclipse.synchronizationTasks` configuration. By using the phased build action API, clients can execute the tasks just before loading the Eclipse model. Also, the task information never needs to leave the Gradle process.

### CODE REVIEW

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+@Incubating
+public interface ExecuteTasksUponEclipseModelLoading {
+    /**
+     * TODO (donat)
+     *
+     * Later, we want to support two use-cases:
+     *  - execute tasks when before loading the Eclipse model
+     *  - execute tasks during Eclipse auto-build
+     *
+     *  Maybe we should refactor the corresponding model builder accept a parameter describing
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java]: No, just have two different commands for that.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import org.gradle.api.Incubating;
+
+/**
+ * TODO (donat) add documentation.
+ *
+ * @since 5.4
+ */
+// TODO (donat) find better naming
+@Incubating
+public interface ExecuteTasksUponEclipseModelLoading {
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java]: RunEclipseSynchronizationTasks

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java [hunk 3]
```diff
@@ -217,6 +223,31 @@ public void jdt(Action<? super EclipseJdt> action) {
...
+    /**
+     * The list of tasks to execute before the Eclipse synchronization starts.
+     *
+     *
+     *
+     * @return the tasks names
+     * @since 5.4
+     */
+    @Incubating
+    public List<String> getSynchronizationTasks() {
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java]: ❌ These shouldn't be Strings, but `Task`s.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        return modelName.equals("org.gradle.tooling.model.eclipse.ExecuteTasksUponEclipseModelLoading");
+    }
+
+    @Override
+    public Object buildAll(String modelName, Project project) {
+        List<String> taskNames = new ArrayList<String>();
+        for (Project p : project.getAllprojects()) {
+            EclipseModel model = p.getExtensions().findByType(EclipseModel.class);
+            if (model != null) {
+                taskNames.addAll(model.getSynchronizationTasks());
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java]: ❌ Task names are not unique, you need to use the path.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            project.task(placeHolderTaskName);
+            taskNames.add(placeHolderTaskName);
+        }
+
+        project.getGradle().getStartParameter().setTaskNames(taskNames);
+        return new EclipseSyncTaskCommandResult(); // TODO (donat) do we need this?
+    }
+
+    private String placeHolderTaskName(Project project, String baseName) {
+        if (project.getTasksByName(baseName, true).isEmpty()) {
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java]: ❌ Never use `getTaskByName(true)`, which configures all subprojects. We don't care about subprojects, we just need a task in this project. Use `findByName`.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        if (taskNames.isEmpty()) {
+            // If no tasks is specified then the default tasks will be executed.
+            // To work around this, we assign a new empty task for execution.
+            String placeHolderTaskName = placeHolderTaskName(project, "nothing");
+            project.task(placeHolderTaskName);
+            taskNames.add(placeHolderTaskName);
+        }
+
+        project.getGradle().getStartParameter().setTaskNames(taskNames);
+        return new EclipseSyncTaskCommandResult(); // TODO (donat) do we need this?
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java]: Nope.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java [hunk 7]
```diff
@@ -67,6 +71,8 @@
 
     private EclipseWtp wtp;
 
+    private List<String> synchronizationTasks = new ArrayList<>();
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java]: Use a `DefaultTaskDependency` as your internal model here, so you can resolve any notation that would be accepted by `Task.dependsOn`

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java [hunk 8]
```diff
@@ -217,6 +223,31 @@ public void jdt(Action<? super EclipseJdt> action) {
...
+
+    /**
+     * Set tasks to be executed before the Eclipse synchronization.
+     *
+     * // TODO (donat) polish documentation.
+     *
+     * @since 5.4
+     */
+    @Incubating
+    public void synchronizationTasks(String... synchronizationTasks) {
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java]: Not just Strings, but any Objects.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java (hunk 1 above)
Comment (donat) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java]: Got that.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java (hunk 2 above)
Comment (donat) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java]: Fixed

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java (hunk 3 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java]: Fixed. In order to parse tasks from String format, I had to inject the current project.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java (hunk 4 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java]: Fixed.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java (hunk 2 above)
Comment (donat) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/model/eclipse/ExecuteTasksUponEclipseModelLoading.java]: Fixed.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java (hunk 5 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java]: Fixed.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java (hunk 6 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/ExecuteEclipseSynchronizationTasks.java]: Removed.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java (hunk 7 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java]: Thanks for the advice, Fixed. (I've learned a lot about the codebase via this PR).

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java (hunk 8 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java]: Fixed.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java [hunk 9]
```diff
@@ -226,14 +231,12 @@ public void jdt(Action<? super EclipseJdt> action) {
...
      *
-     *
-     *
      * @return the tasks names
      * @since 5.4
      */
     @Incubating
-    public List<String> getSynchronizationTasks() {
-        return this.synchronizationTasks;
+    public DefaultTaskDependency getSynchronizationTasks() {
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java]: Should not expose internal type (just return an immutable set)

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunEclipseSynchronizationTasksBuilder.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+        if (taskPaths.isEmpty()) {
+            // If no tasks is specified then the default tasks will be executed.
+            // To work around this, we assign a new empty task for execution.
+            String placeHolderTaskName = placeHolderTaskName(project, "nothing");
+            project.task(placeHolderTaskName);
+            taskPaths.add(placeHolderTaskName);
+        }
+
+        project.getGradle().getStartParameter().setTaskNames(taskPaths);
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunEclipseSynchronizationTasksBuilder.java]: Should respect any tasks that were there before

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java (hunk 9 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseModel.java]: Shoot, I forgot to add a todo there :) Fixed.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunEclipseSynchronizationTasksBuilder.java (hunk 10 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunEclipseSynchronizationTasksBuilder.java]: Fixed and added a test for it.

**Pull request merged**